File: 0217-contains-duplicate/0217-contains-duplicate.cpp
class Solution {
public:
    bool containsDuplicate(vector<int>& nums) {
        // map<int,int> mpp;
        // for(int i=0;i<nums.size();i++){
        //     mpp[nums[i]]++;
        // }
        // for(auto it: mpp){
        //     if(it.second >= 2){
        //         return true;
        //     }
        // }
        // return false;


        sort(nums.begin(), nums.end());
        for(int i=1;i<nums.size();i++){
            if(nums[i] == nums[i-1]){
                return true;
            }
        }
        return false;

        // unordered_set<int> seen;
        // for(auto num: nums){
        //     if(seen.count(num)) return true;
        //     seen.insert(num);
        // }
        // return false;
    }
};

File: 0217-contains-duplicate/0217-contains-duplicate.py
class Solution(object):
    def containsDuplicate(self, nums):
        """
        :type nums: List[int]
        :rtype: bool
        """
        
        ### brute force
        #TC O(n2)
        #SC O(1)

        # for i in range(len(nums)):
        #     for j in range(i+1,len(nums)):
        #         if nums[i]==nums[j]:
        #             return True

        
        # return False

        ###sorting
        ## TC O(nlogn)
        ## SC O(1) 

        # nums.sort()

        # for i in range(1,len(nums)):
        #     if nums[i]==nums[i-1]:
        #         return True

        # return False

        ### HashSet
        ## TC : O(n)
        ## SC: O(n)

        # seen=set()
        # for num in nums:
        #     if num in seen:
        #         return True
        #     seen.add(num)

        # return False

        ### HashSet Length
        ## TC : O(n)
        ## SC: O(n)

        return len(set(nums)) < len(nums)




        ## using hashset will be trade off between using sorting as in hashing
        ## you will sacrifice space and reduce time complexity